xdsl/transforms/eqsat_serialize_egraph.py

```diff
@@ -0,0 +1,75 @@
+import json
+from collections import defaultdict
+from typing import Any, cast
+
+from xdsl.context import Context
+from xdsl.dialects import builtin, eqsat
+from xdsl.ir import BlockArgument, Operation
+from xdsl.passes import ModulePass
+
+
+class IDGenerator:
+    def __init__(self, prefix: str):
+        self.prefix = prefix
+        self.counter = 0
+
+    def __call__(self):
+        self.counter += 1
+        return f"{self.prefix}{self.counter}"
+
+
+def serialize_to_egraph(mod: builtin.ModuleOp):
+    enode_to_id: defaultdict[Operation | BlockArgument, str] = defaultdict(
+        IDGenerator("enode_")
+    )
+    eclass_to_id: defaultdict[eqsat.EClassOp, str] = defaultdict(IDGenerator("eclass_"))
+    nodes: dict[str, dict[str, str | list[str]]] = dict()
+    for op in mod.walk(reverse=True):
+        if isinstance(op, eqsat.EClassOp):
+            for operand in op.operands:
+                if isinstance(operand, BlockArgument):
+                    nodes[enode_to_id[operand]] = {
+                        "op": f"arg {operand.index}",
+                        "eclass": eclass_to_id[op],
+                        "children": [],
+                    }
+            continue
+        children: list[Any] = []
+        eclass_id = None
+        for res in op.results:
+            for use in res.uses:
+                if isinstance(use.operation, eqsat.EClassOp):
+                    assert len(op.results) == 1, (
+                        "Only single result operations are supported"
+                    )
+                    assert len(res.uses) == 1, (
+                        "Only single use operations are supported"
+                    )
+                    eclass_id = eclass_to_id[use.operation]
+        if eclass_id is None:
+            continue
+        for operand in op.operands:
+            if isinstance(operand.owner, eqsat.EClassOp):
+                firstoperand = operand.owner.operands[0]
+                if isinstance(firstoperand.owner, Operation):
+                    canonical_enode = firstoperand.owner
+                else:
+                    canonical_enode = cast(BlockArgument, firstoperand)
+                children.append(enode_to_id[canonical_enode])
+        if len(op.operands) > 0:
+            name = op.name
+        else:
+            name = str(op).split("=")[1].split(":")[0].strip()
```

Comment:
Neat, but worth a comment
```suggestion
        if op.operands:
            name = op.name
        else:
            name = str(op).split("=")[1].split(":")[0].strip()
```

Reply:
Upon further thought, the string representation of an operation doesn't always include a colon followed by the type. Perhaps we can just keep the type in the name as well?